File: core/WordBreakII/WordBreakII.py
from typing import List
from collections import deque


def wordBreak(s: str, wordDict: List[str]) -> List[str]:
    res = []
    def dfs(sub_str, remain):
        if sub_str == "":
            res.append(" ".join(remain))
            return

        length = len(sub_str)
        for i in range(length - 1, -1, -1):
            if sub_str[i:] in wordDict:
                remain.appendleft(sub_str[i:])
                dfs(sub_str[:i], remain)
                remain.popleft()
    dfs(s, deque())
    return res


if __name__ == '__main__':
    print(wordBreak("catsanddog", ["cat","cats","and","sand","dog"]))

File: core/WordBreakII/WordBreakII.java
package com.dbc;

import java.util.*;

public class WordBreakII {
    private void dfs(String str, Deque<String> queue, List<String> res, List<String> wordDict) {
        if (str.equals("")) {
            res.add(String.join(" ", queue));
        }
        for (int i = str.length() - 1; i >= 0; i--) {
            String subStr = str.substring(i);
            if (wordDict.contains(subStr)) {
                queue.addFirst(subStr);
                dfs(str.substring(0, i), queue, res, wordDict);
                queue.pollFirst();
            }
        }
    }

    public List<String> wordBreak(String s, List<String> wordDict) {
        List<String> res = new ArrayList<>();
        dfs(s, new ArrayDeque<>(), res, wordDict);
        return res;
    }
}
